### DISCUSSION TITLE
Use functions from parent file in child file

### DISCUSSION THREAD

**Discussion by cocoveal** [typst/typst#4809] [Q&A] [answered]
I am trying to have my own custom document-wide functions, which would sit in a parent file, which includes the children files that need them. The reason why is because my parent file would be a template. This would look something like this:

parent.typ:
```typst
#let parskip() = {parbreak(); v(1em)}
#include "child.typ"
//or
#import "child.typ"
```

child.typ:
```typst
content
#parskip()
content
```

I am new to Typst and I've tried looking for solutions online, but haven't found anything. I've tried state and querying, but couldn't get it to work. I've seen some templates that provide functions, but in their example files, they don't do anything, it just works out of the box, like described above, so I'm confused. Am I forgetting something or doing something wrong?

**Comment by benediktschfr** [comment]
Did you try to import the parent file in the child file and not the other way around?

**Reply by cocoveal** [reply]
Hey Benedikt, 
no I have not. The parent file is holds all the formatting, etc. The children files are supposed to just have content on them. All the styling, formatting, etc. is done by the parent file, so the children files are not clustered with code, etc. 

I am writing a scientific paper, therefore I want one main file (parent) that contains all the template stuff and just includes the content files (children) and does not have any content itself.

**Reply by benediktschfr** [reply]
That's how it's supposed to be. Have a look at the "How to make a template" section in the official documentation. You can define as many functions as you want in the parent file and import that file in the child file (the import command has to be in the child file). Then you can access all your functions.

You did it the other way around, as far as I understand. You included the child file on the parent file, which does basically nothing.

**Reply by cocoveal** [reply]
I'm sorry, I think I misspoke. I understand all of that, but my use case is different.

What I have is a template. This template is my main file, that centralizes everything for me. Then my chapters that I write, I separate in different files to make everything cleaner. I then import/inject these into the main file. I have issues accessing the functions in all the children files. I don't want to have one big file containing all my content, but I don't want to redefine the same function in every single content file. 

I use this approach so my content files are clean and also to avoid any issues with bibliographies and the table of contents. So, is there a workaround that could help me, or is this something that is impossible with the typst compiler?

**Reply by laurmaedje** [reply, marked as answer]
The usual setup is to have a main file with an everything `show:` rule that applies your template and `include`s for your chapters, chapter files, and a template file, which you import in all three. So something like this:

```typ
// main.typ
#import "template.typ": template

#show: template.with(
  title: [My paper],
  ..
)

#include "chapter1.typ"
#include "chapter2.typ"
...
```

```typ
// chapter1.typ
#import "template.typ": *

This is some text. I can use a
#utility() here.
```

```typ
// template.typ
#let template(.., body) = {
  // Defines all the styling
  body
}

#let utility(..) = {
  // Helper function
}
```

**Reply by cocoveal** [reply]
Hey,

thanks, yeah, I guess that would work. A little bit tedious importing the template functions in every child file, but it's less work than the other options.

**Reply by laurmaedje** [reply]
I can see how one might find it a bit tedious, but ultimately I think it's a fair price to pay in exchange for each file being well-defined by itself. If function definitions could somehow come from the outside, the module system would become complex and hard to understand.

**Reply by cocoveal** [reply]
I guess it only makes sense. It is the same in programming in general. I guess it's just wishful thinking because of all of the other cool stuff typst does compared to other Markup languages when it comes to scripting.
